### PR TITLE
Feature/155 independent api types in backend

### LINKED ISSUE #155: Types returned by journeys backend should be independent of API used
The "journey"/"trip" module in backend depends on a 3rd party library that's being used to plan a journey. This 3rd party library returns it's own types. This types should not be transferred to frontend - frontend should be independent of any 3rd party library which is likely to change over time. (For now _vrr_client_ is being used as 3rd party library.)
To achieve this, a custom "type layer" has to be defined in backend that defines all types we want to use within our app. 3rd party specific types are being converted to our custom types.

Steps:
- [x] Define custom types/dtos for ~Stops~ Locations
- [x] Implement converter for ~Stops~ Locations (vrr_client type -> custom type)
- [x] Define custom types/dtos for Trips
- [x] Implement converter/mapper for Trips (vrr_client type -> custom type)
- [x] Re-Generate frontend API files
- [x] Adjust dependent frontend parts (e.g. journeyDemo)

### PR DESCRIPTION
**Pull request by JaJako**
### Description

Introduces custom types to uncouple the types used in backend/frontend from 3rd party library types. Resolves #155. 

## Checklist

- [x] My code follows the style guidelines
- [x] I have documented my code
- [x] My changes generate no new warnings
- [x] I have linked the Issue
- [x] The Pull Request covers all the Checkpoints from the tagged Issue
- [x] I have noted the hours of work i have put in the issue

## I have edited following Parts of the project

- [x] Frontend
- [x] Backend
- [ ] Github
- [ ] Root

## If tests are required did you make them?

- [x] No tests required
- [ ] Yes
- [ ] No

### CODE REVIEW

**Comment by JaJako** [comment]
@dkrakau I've created this PR draft, so that our progress so far is transparent/visible to the others.

**Comment by timvahlbrock** [comment]
@dkrakau How far are you with this PR? If this isn't close to ready we need to implement the route planning functionality with the old types first.

**Comment by dkrakau** [comment]
@timvahlbrock This PR should be done now.

**Comment by Paulomart** [comment]
@dkrakau Please resolve conflicts before requesting a review

**Comment by JaJako** [comment]
@timvahlbrock @Paulomart This is ready for review now.

File: backend/src/journey/entity/journey.entity.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  legs?: Leg[];
+
+  constructor(legs: Leg[]) {
+    this.legs = legs;
+  }
+
+}
+
+
+/* export interface TRIPSchema {
```
Comment (Paulomart) [inline backend/src/journey/entity/journey.entity.ts:19]: Remove commented code

File: backend/src/journey/entity/journeyLocation.entity.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    this.departure = departure;
+  }
+
+}
+
+/* 
+export interface JourneyLocationElement {
+  assignedStops?: AssignedLocation[];
+  buildingNumber?: string;
+  coord?: Array<any[] | CoordClass | number>;
```
Comment (Paulomart) [inline backend/src/journey/entity/journeyLocation.entity.ts:35]: Same here

File: backend/src/journey/entity/leg.entity.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+
+/* 
+export interface Leg {
+    coords?: Array<Array<any[] | CoordClass | number>>;
+    destination?: JourneyLocationElement;
+    distance?: number;
+    duration?: number;
+    elevationSummary?: ElevationSummary;
+    fare?: LegFare;
```
Comment (Paulomart) [inline backend/src/journey/entity/leg.entity.ts:49]: Same here

File: backend/src/journey/entity/legDetails.entity.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    destination?: JourneyLocationElement;
+    distance?: number;
+    duration?: number;
+    elevationSummary?: ElevationSummary;
+    fare?: LegFare;
+    footPathInfo?: Array<any[] | boolean | FootPathInfoClass | number | number | null | string>;
+    hints?: Info[];
+    infos?: Info[];
+    interchange?: Interchange;
+    isRealtimeControlled?: boolean;
```
Comment (Paulomart) [inline backend/src/journey/entity/legDetails.entity.ts:58]: Same here

File: backend/src/journey/entity/transportation.entity.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+export interface Transportation {
+    allCoords?: Array<Array<Array<any[] | CoordClass | number>>>;
+    assignedStop?: string;
+    assignedStopId?: string;
+    coords?: Array<Array<Array<any[] | CoordClass | number>>>;
+    description?: string;
+    destination?: Location;
+    disassembledName?: string;
+    id?: string;
+    index?: string;
```
Comment (Paulomart) [inline backend/src/journey/entity/transportation.entity.ts:39]: Same here

File: backend/src/journey/entity/trip.entity.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    this.status = status;
+    this.trainNumber = trainNumber;
+  }
+
+}
+
+/* export interface TransportationTrip {
+  arrivalTimePlannedJourneyDestination?: string;
+  departureTimePlannedJourneyOrigin?: string;
+  status?: Status;
```
Comment (Paulomart) [inline backend/src/journey/entity/trip.entity.ts:42]: Same here

**Review by Paulomart — CHANGES_REQUESTED** [review]
Schaue mir den Code nachher nochmal genauer an

File: backend/src/journey/controller/journey.controller.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  @ApiOkResponse({
+    description: "Journeys.",
+    type: [Journey]
+  })
+  @ApiBadRequestResponse({
+    description: "Bad request.",
+    type: BadRequest
+  })
+  public queryJourney(@Body() requestBody: JourneyRequestDto): Promise<Journey[]> {
+    return this.journeyService.queryJourney(requestBody.originId, requestBody.destinationId/*, requestBody.departure*/);
```
Comment (Paulomart) [inline backend/src/journey/controller/journey.controller.ts:26]: The departure time is missing here. Any reason for that?

File: backend/src/journey/entity/legDetails.entity.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
+import { ApiProperty } from "@nestjs/swagger";
+
+export class LegDetails {
+
+  @ApiProperty({
+    description: "Distance.",
+    type: Number
+  })
+  distance?: number;
```
Comment (Paulomart) [inline backend/src/journey/entity/legDetails.entity.ts:9]: distanceInXXX

File: backend/src/journey/entity/legDetails.entity.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+  @ApiProperty({
+    description: "Duration.",
+    type: Number
+  })
+  duration?: number;
```
Comment (Paulomart) [inline backend/src/journey/entity/legDetails.entity.ts:15]: durationInXXX

File: backend/src/journey/entity/legDetails.entity.ts [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+  @ApiProperty({
+    description: "Leg information.",
+    type: String
+  })
+  infos?: string; // Info[]
```
Comment (Paulomart) [inline backend/src/journey/entity/legDetails.entity.ts:21]: What does this mean? If we dont have a better name for that, remove it and we add it back once we notice that we need it.

File: backend/src/journey/entity/legDetails.entity.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+  @ApiProperty({
+    description: "Leg hints.",
+    type: String
+  })
+  hints?: string; // Info[]
```
Comment (Paulomart) [inline backend/src/journey/entity/legDetails.entity.ts:27]: I dont like the typing here either. I assume the vrr api returns a string?

File: backend/src/journey/entity/time.entity.ts [hunk 12]
```diff
@@ -0,0 +1,30 @@
+import { ApiProperty } from "@nestjs/swagger";
+
+export class Time {
```
Comment (Paulomart) [inline backend/src/journey/entity/time.entity.ts:3]: I guess this should mean when a event occurs?

File: backend/src/journey/entity/time.entity.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    description: "Time base.",
+    type: String
+  })
+  timeBaseTimetable?: string;
+
+  @ApiProperty({
+    description: "Time estimated.",
+    type: String
+  })
+  timeEstimated?: string;
```
Comment (Paulomart) [inline backend/src/journey/entity/time.entity.ts:15]: drop the common prefix: estimated

File: backend/src/journey/entity/time.entity.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+  @ApiProperty({
+    description: "Time planned.",
+    type: String
+  })
+  timePlanned?: string;
```
Comment (Paulomart) [inline backend/src/journey/entity/time.entity.ts:21]: drop the common prefix: planned

File: backend/src/journey/entity/time.entity.ts (hunk 12 above)
Comment (Paulomart) [reply backend/src/journey/entity/time.entity.ts:3]: Also: Why is time stored as a string?

File: backend/src/journey/entity/transportation.entity.ts [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    required: true
+  })
+  name?: string;
+
+  @ApiProperty({
+    description: "Trips depending on this transportation.",
+    type: [Trip],
+    required: true
+  })
+  trips!: Trip[];
```
Comment (Paulomart) [inline backend/src/journey/entity/transportation.entity.ts:19]: Is this information not already contained in `leg`?

**Review by Paulomart — APPROVED** [review]

**Review by Paulomart — CHANGES_REQUESTED** [review]
wrong button...

File: backend/src/journey/controller/journey.controller.ts (hunk 7 above)
Comment (JaJako) [reply backend/src/journey/controller/journey.controller.ts:26]: Departure time and "asArrival" is passed to the service now. But vrr_client is not supporting those two parameters yet. Should we create a new issue for that?

Comment (Paulomart) [reply backend/src/journey/controller/journey.controller.ts:26]: Okay I see know. Maybe the service should ignore the parameter than and I will create an issue to implement that.

Comment (Paulomart) [reply backend/src/journey/controller/journey.controller.ts:26]: 267

File: backend/src/journey/entity/time.entity.ts [hunk 16]
```diff
@@ -0,0 +1,30 @@
+import { ApiProperty } from "@nestjs/swagger";
+
+export class Time {
+
+  @ApiProperty({
+    description: "Time base.",
+    type: String
+  })
+  timeBaseTimetable?: string;
```
[the remark this answers is not in the document]
Comment (JaJako) [reply backend/src/journey/entity/time.entity.ts:9]: As there seems to be no infos supplied by VRR for this field, we removed it.

File: backend/src/journey/entity/transportation.entity.ts (hunk 15 above)
Comment (JaJako) [reply backend/src/journey/entity/transportation.entity.ts:19]: The trip information is about the transportation itself. So it's not about the leg of a journey itself but about the transportation that is used by the leg.

File: backend/src/journey/entity/legDetails.entity.ts [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+  @ApiProperty({
+    description: "Leg real time status.",
+    type: String
+  })
+  realtimeStatus?: string; // RealtimeTripStatus
```
[the remark this answers is not in the document]
Comment (JaJako) [reply backend/src/journey/entity/legDetails.entity.ts:33]: Vrr api uses an enum with string values here. We would think a string is enough for now?

File: backend/src/journey/entity/transportation.entity.ts (hunk 15 above)
Comment (Paulomart) [reply backend/src/journey/entity/transportation.entity.ts:19]: Kurz mal auf deutsch: Also sind Trips Fahrten die ein Fahrzeug (oder eine Line) macht?

Comment (JaJako) [reply backend/src/journey/entity/transportation.entity.ts:19]: > Kurz mal auf deutsch: Also sind Trips Fahrten die ein Fahrzeug (oder eine Line) macht?

Genau, das gibt an, was das Fahrzeug/Transportmittel aus einem Leg so tut.

File: backend/src/journey/entity/legDetails.entity.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (Paulomart) [reply backend/src/journey/entity/legDetails.entity.ts:33]: If we need it we should do our own enum here. We ne dont need it we should drop the field.

File: backend/src/journey/entity/transportation.entity.ts (hunk 15 above)
Comment (Paulomart) [reply backend/src/journey/entity/transportation.entity.ts:19]: Okay dann eine Idee für den Kommentar: 

> Trips that are executed by this transportation.

Transportation hängt dann aber nicht von Leg ab.

File: backend/src/journey/entity/trip.entity.ts [hunk 18]
```diff
@@ -0,0 +1,46 @@
+import { ApiProperty } from "@nestjs/swagger";
+
+export class Trip {
+
+  @ApiProperty({
+    description: "Planned arrival time at destination.",
+    type: String
+  })
+  arrivalTimePlannedJourneyDestination?: string;
```
Comment (Paulomart) [inline backend/src/journey/entity/trip.entity.ts:9]: Ah das hier hatte mich verwirrt. @JaJako hier hängt Trip von dem geplanten Journey ab.

Comment (JaJako) [reply backend/src/journey/entity/trip.entity.ts:9]: Der Name (übernommen von VRR) ist da vielleicht verwirrend, aber die "Journey" wird sich hier auf das Fahrzeug beziehen (und welche Reise dieses unternimmt).

File: backend/src/journey/entity/legDetails.entity.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (JaJako) [reply backend/src/journey/entity/legDetails.entity.ts]: We faced some problems trying to map the vrr enum to a custom enum. Writing a big switch statement for all values vrr is offering to set our custom values seems a bit too much and would need to be maintained with every changes that is made to vrr enum or our custom enum. Approaches to build a programmatic mapping did not work or were confusing. Created a new issue to look at using enums for later (now sticking to strings).

[the remark this answers is not in the document]
Comment (JaJako) [reply backend/src/journey/entity/legDetails.entity.ts]: #270

File: backend/src/journey/entity/trip.entity.ts (hunk 18 above)
Comment (Paulomart) [reply backend/src/journey/entity/trip.entity.ts:9]: Okay. Falls wir merken, dass das Feld komisch ist, können wir ja immer noch iterieren 👍

**Review by Paulomart — CHANGES_REQUESTED** [review]

**Review by Paulomart — APPROVED** [review]

**Pull request merged**